common/src/main/java/com/scottlogic/deg/generator/fieldspecs/whitelist/FrequencyDistributedSet.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2019 Scott Logic Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.scottlogic.deg.generator.fieldspecs.whitelist;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class FrequencyDistributedSet<T> implements DistributedSet<T> {
+
+    private static final FrequencyDistributedSet<?> EMPTY = new FrequencyDistributedSet<>(Collections.emptySet());
+
+    private final Set<WeightedElement<T>> underlyingWeights;
+
+    private final List<WeightedElement<T>> underlyingCumulativeWeights;
+
+    public FrequencyDistributedSet(final Set<WeightedElement<T>> underlyingWeights) {
+        if (underlyingWeights.contains(null)) {
+            throw new IllegalArgumentException("FrequencyDistributedSet should not contain null elements");
+        }
+
+        Set<WeightedElement<T>> normalisedWeights = normalise(underlyingWeights);
+        this.underlyingWeights = Collections.unmodifiableSet(normalisedWeights);
+        this.underlyingCumulativeWeights = cumulative(normalisedWeights);
+    }
+
+    private static <T> Set<WeightedElement<T>> normalise(final Set<WeightedElement<T>> denormalised) {
```

Comment:
can we order the methods interface public then private, then other publics like tostring

Reply:
I disagree with this - I find it more sensible to group methods based on their usage. Public methods should read intuitively, with their associated private methods declaration in close proximity to their usage.